Flaky test `test_edgeql_ddl_migration_sdl_01`

See also https://github.com/edgedb/edgedb/actions/runs/10764934440/attempts/1:

```
================================================================================
FAIL: test_edgeql_ddl_migration_sdl_01 (test_edgeql_ddl.TestEdgeQLDDL.test_edgeql_ddl_migration_sdl_01) {serialization='json'}
--------------------------------------------------------------------------------
Py_HashSecret: 78aa45e26491b051cf8a249c34617f8ac7824fc27ae66f84
random.seed(): 4ee9b0357b7f365e
--------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/aarch64-linux-gnu/edgedb-server-6-dev8784/lib/python3.12/site-packages/edb/testbase/server.py", line 209, in wrapper
    self.loop.run_until_complete(
  File "/usr/lib/aarch64-linux-gnu/edgedb-server-6-dev8784/lib/python3.12/asyncio/base_events.py", line 685, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/usr/share/edgedb-server-6-dev8784/tests/test_edgeql_ddl.py", line 13698, in test_edgeql_ddl_migration_sdl_01
    await self.assert_query_result(
  File "/usr/lib/aarch64-linux-gnu/edgedb-server-6-dev8784/lib/python3.12/site-packages/edb/testbase/server.py", line 1123, in assert_query_result
    assert_data_shape.assert_data_shape(
  File "/usr/lib/aarch64-linux-gnu/edgedb-server-6-dev8784/lib/python3.12/site-packages/edb/common/assert_data_shape.py", line 375, in assert_data_shape
    return _assert_generic_shape((), data, shape)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/aarch64-linux-gnu/edgedb-server-6-dev8784/lib/python3.12/site-packages/edb/common/assert_data_shape.py", line 331, in _assert_generic_shape
    return _assert_list_shape(path, data, shape)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/aarch64-linux-gnu/edgedb-server-6-dev8784/lib/python3.12/site-packages/edb/common/assert_data_shape.py", line 156, in _assert_list_shape
    _assert_generic_shape(
  File "/usr/lib/aarch64-linux-gnu/edgedb-server-6-dev8784/lib/python3.12/site-packages/edb/common/assert_data_shape.py", line 335, in _assert_generic_shape
    return _assert_dict_shape(path, data, shape)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/aarch64-linux-gnu/edgedb-server-6-dev8784/lib/python3.12/site-packages/edb/common/assert_data_shape.py", line 115, in _assert_dict_shape
    _assert_generic_shape(path + (f'["{sk}"]',), data[sk], sv)
  File "/usr/lib/aarch64-linux-gnu/edgedb-server-6-dev8784/lib/python3.12/site-packages/edb/common/assert_data_shape.py", line 353, in _assert_generic_shape
    fail(
  File "/usr/lib/aarch64-linux-gnu/edgedb-server-6-dev8784/lib/python3.12/site-packages/edb/common/assert_data_shape.py", line 76, in fail
    base_fail(f'{msg}\nshape: {shape!r}\ndata: {data!r}')
AssertionError: data shape differs: 'module default {\n    type B;\n};' != 'module default {\n    type A;\n};' PATH: [0]["sdl"]
shape: [{'sdl': 'module default {\n    type A;\n};'}, {'sdl': 'module default {\n    type A;\n    type B {\n        property n: std::int64;\n    };\n};'}, {'sdl': 'module default {\n    type A {\n        link b: default::B;\n    };\n    type B {\n        property n: std::int64;\n    };\n};'}, {'sdl': 'module default {\n    type A {\n        link b: default::B {\n            property n: std::int64;\n        };\n    };\n    type B {\n        property n: std::int64;\n    };\n};'}, {'sdl': 'module default {\n    type A {\n        link b: default::B;\n    };\n    type B {\n        property n: std::int64;\n    };\n};'}, {'sdl': 'module default {\n    type A {\n        link b: default::B;\n    };\n    type B;\n};'}, {'sdl': 'module default {\n    type B;\n};'}]
data: [{'sdl': 'module default {\n    type B;\n};'}, {'sdl': 'module default {\n    type A;\n};'}, {'sdl': 'module default {\n    type A;\n    type B {\n        property n: std::int64;\n    };\n};'}, {'sdl': 'module default {\n    type A {\n        link b: default::B;\n    };\n    type B {\n        property n: std::int64;\n    };\n};'}, {'sdl': 'module default {\n    type A {\n        link b: default::B {\n            property n: std::int64;\n        };\n    };\n    type B {\n        property n: std::int64;\n    };\n};'}, {'sdl': 'module default {\n    type A {\n        link b: default::B;\n    };\n    type B {\n        property n: std::int64;\n    };\n};'}, {'sdl': 'module default {\n    type A {\n        link b: default::B;\n    };\n    type B;\n};'}]
```

--- a/tests/test_edgeql_ddl.py
+++ b/tests/test_edgeql_ddl.py
@@ -13698,7 +13698,7 @@ type default::Foo {
 
         # Migrations implicitly ordered based on parent_id.
         # Fetch the migrations and sort them here.
-        migrations = json.loads(await self.con._fetchall_json(
+        migrations = json.loads(await self.con.query_json(
             'select schema::Migration { id, parent_ids := .parents.id, sdl }'
         ))
 

--- a/tests/test_edgeql_ddl.py
+++ b/tests/test_edgeql_ddl.py
@@ -18,6 +18,7 @@
 
 import asyncio
 import decimal
+import json
 import os
 import re
 import textwrap
@@ -13663,7 +13664,6 @@ type default::Foo {
                     DROP FUNCTION foo___1(a: int64);
                 ''')
 
-    @test.skip('Too flaky; see #7729')
     async def test_edgeql_ddl_migration_sdl_01(self):
         await self.con.execute('''
             CONFIGURE SESSION SET store_migration_sdl :=
@@ -13696,81 +13696,104 @@ type default::Foo {
             drop type A;
         ''')
 
-        await self.assert_query_result(
-            'select schema::Migration { sdl }',
+        # Migrations implicitly ordered based on parent_id.
+        # Fetch the migrations and sort them here.
+        migrations = json.loads(await self.con._fetchall_json(
+            'select schema::Migration { id, parent_ids := .parents.id, sdl }'
+        ))
+
+        # Find migrations with no parents.
+        sdl = [
+            migration['sdl']
+            for migration in migrations
+            if not migration['parent_ids']
+        ]
+
+        # Repeatedly find descendents until no more can be found.
+        prev_ids = [
+            migration['id']
+            for migration in migrations
+            if not migration['parent_ids']
+        ]
+        while prev_ids:
+            sdl.extend(
+                migration['sdl']
+                for migration in migrations
+                if any(
+                    parent_id in prev_ids
+                    for parent_id in migration['parent_ids']
+                )
+            )
+            prev_ids = [
+                migration['id']
+                for migration in migrations
+                if any(
+                    parent_id in prev_ids
+                    for parent_id in migration['parent_ids']
+                )
+            ]
+
+        self.assert_data_shape(
+            sdl,
             [
-                {
-                    'sdl': (
-                        'module default {\n'
-                        '    type A;\n'
-                        '};'
-                    ),
-                },
-                {
-                    'sdl': (
-                        'module default {\n'
-                        '    type A;\n'
-                        '    type B {\n'
-                        '        property n: std::int64;\n'
-                        '    };\n'
-                        '};'
-                    ),
-                },
-                {
-                    'sdl': (
-                        'module default {\n'
-                        '    type A {\n'
-                        '        link b: default::B;\n'
-                        '    };\n'
-                        '    type B {\n'
-                        '        property n: std::int64;\n'
-                        '    };\n'
-                        '};'
-                    ),
-                },
-                {
-                    'sdl': (
-                        'module default {\n'
-                        '    type A {\n'
-                        '        link b: default::B {\n'
-                        '            property n: std::int64;\n'
-                        '        };\n'
-                        '    };\n'
-                        '    type B {\n'
-                        '        property n: std::int64;\n'
-                        '    };\n'
-                        '};'
-                    ),
-                },
-                {
-                    'sdl': (
-                        'module default {\n'
-                        '    type A {\n'
-                        '        link b: default::B;\n'
-                        '    };\n'
-                        '    type B {\n'
-                        '        property n: std::int64;\n'
-                        '    };\n'
-                        '};'
-                    ),
-                },
-                {
-                    'sdl': (
-                        'module default {\n'
-                        '    type A {\n'
-                        '        link b: default::B;\n'
-                        '    };\n'
-                        '    type B;\n'
-                        '};'
-                    ),
-                },
-                {
-                    'sdl': (
-                        'module default {\n'
-                        '    type B;\n'
-                        '};'
-                    ),
-                },
+                (
+                    'module default {\n'
+                    '    type A;\n'
+                    '};'
+                ),
+                (
+                    'module default {\n'
+                    '    type A;\n'
+                    '    type B {\n'
+                    '        property n: std::int64;\n'
+                    '    };\n'
+                    '};'
+                ),
+                (
+                    'module default {\n'
+                    '    type A {\n'
+                    '        link b: default::B;\n'
+                    '    };\n'
+                    '    type B {\n'
+                    '        property n: std::int64;\n'
+                    '    };\n'
+                    '};'
+                ),
+                (
+                    'module default {\n'
+                    '    type A {\n'
+                    '        link b: default::B {\n'
+                    '            property n: std::int64;\n'
+                    '        };\n'
+                    '    };\n'
+                    '    type B {\n'
+                    '        property n: std::int64;\n'
+                    '    };\n'
+                    '};'
+                ),
+                (
+                    'module default {\n'
+                    '    type A {\n'
+                    '        link b: default::B;\n'
+                    '    };\n'
+                    '    type B {\n'
+                    '        property n: std::int64;\n'
+                    '    };\n'
+                    '};'
+                ),
+                (
+                    'module default {\n'
+                    '    type A {\n'
+                    '        link b: default::B;\n'
+                    '    };\n'
+                    '    type B;\n'
+                    '};'
+                ),
+                (
+                    'module default {\n'
+                    '    type B;\n'
+                    '};'
+                ),
             ]
         )
 

